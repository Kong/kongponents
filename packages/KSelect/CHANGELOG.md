# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 6.0.1-alpha.1 (2021-11-16)

**Note:** Version bump only for package @kongponents/kselect





## 6.0.1-alpha.0 (2021-11-16)

**Note:** Version bump only for package @kongponents/kselect





## 5.0.4-alpha.0 (2021-11-16)

**Note:** Version bump only for package @kongponents/kselect
