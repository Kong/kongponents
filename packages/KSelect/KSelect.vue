<template>
  <div
    :style="widthStyle"
    class="k-select"
  >
    <KLabel
      v-if="label"
      :for="selectId"
      v-bind="labelAttributes"
      data-testid="k-select-label">{{ label }}</KLabel>
    <div
      :id="selectId"
      data-testid="k-select-selected-item">
      <div
        v-if="selectedItem && appearance === 'dropdown'"
        class="k-select-item-selection px-3 py-1">
        <div
          class="selected-item-label">{{ selectedItem.label }}</div>
        <button
          class="clear-selection-icon cursor-pointer non-visual-button"
          @click="clearSelection"
          @keyup.enter="clearSelection">
          <KIcon
            color="var(--blue-200)"
            icon="clear"
          />
        </button>
      </div>
      <KToggle v-slot="{toggle, isToggled}">
        <KPop
          v-bind="boundKPopAttributes"
          :on-popover-click="() => {
            toggle()
            return isToggled
          }"
          :position-fixed="positionFixed"
          :test-mode="testMode"
          :target="`[id='${selectInputId}']`"
          @opened="() => {
            filterStr = ''
            toggle()
          }"
          @closed="() => {
            if (selectedItem && appearance === 'select') {
              filterStr = selectedItem.label
            }
            if (isToggled) {
              toggle()
            }
          }"
        >
          <div
            v-if="appearance === 'button'"
            :id="selectInputId"
            class="k-select-button"
            data-testid="k-select-input"
            style="position: relative;"
            role="listbox">
            <KButton
              :style="widthStyle"
              :id="selectTextId"
              :is-open="isToggled"
              :is-rounded="false"
              v-bind="$attrs"
              appearance="btn-link"
              @keyup="triggerFocus(isToggled)">{{ selectButtonText }}</KButton>
          </div>
          <div
            v-else
            :id="selectInputId"
            :class="{ 'k-select-input': appearance === 'select'}"
            data-testid="k-select-input"
            style="position: relative;"
            role="listbox">
            <KIcon
              v-if="appearance === 'select'"
              icon="chevronDown"
              color="var(--grey-500)"
              size="15" />
            <KInput
              :is-open="isToggled"
              :id="selectTextId"
              v-bind="$attrs"
              v-model="filterStr"
              :placeholder="placeholderText"
              class="k-select-input"
              @keyup="triggerFocus(isToggled)" />
          </div>
          <template v-slot:content>
            <ul class="k-select-list ma-0 pa-0">
              <slot
                :items="items"
                :is-open="isToggled"
                name="items"
              >
                <KSelectItem
                  v-for="item in filteredItems"
                  :key="item.key"
                  :item="item"
                  @selected="handleItemSelect">
                  <template v-slot:content>
                    <slot
                      :item="item"
                      name="item-template"
                    />
                  </template>
                </KSelectItem>
                <KSelectItem
                  v-if="!filteredItems.length"
                  key="k-select-empty-state"
                  :item="{ label: 'No results', value: 'no_results' }"
                  class="k-select-empty-item"/>
              </slot>
            </ul>
          </template>
        </KPop>
      </KToggle>
    </div>
  </div>
</template>

<script>
import { uuid } from 'vue-uuid'
import KButton from '@kongponents/kbutton/KButton.vue'
import KIcon from '@kongponents/kicon/KIcon.vue'
import KInput from '@kongponents/kinput/KInput.vue'
import KLabel from '@kongponents/klabel/KLabel.vue'
import KPop from '@kongponents/kpop/KPop.vue'
import KToggle from '@kongponents/ktoggle/KToggle'
import KSelectItem from './KSelectItem.vue'

const defaultKPopAttributes = {
  hideCaret: true,
  popoverClasses: 'k-select-popover mt-0',
  popoverTimeout: 0,
  placement: 'bottomStart'
}

export default {
  name: 'KSelect',
  components: { KButton, KIcon, KInput, KLabel, KPop, KSelectItem, KToggle },
  inheritAttrs: false,
  props: {
    value: {
      type: [String, Number],
      default: ''
    },
    kpopAttributes: {
      type: Object,
      default: () => ({
        popoverClasses: ''
      })
    },
    label: {
      type: String,
      default: ''
    },
    labelAttributes: {
      type: Object,
      default: () => ({})
    },
    /**
     * The width of the select and popover's min-width
     */
    width: {
      type: String,
      default: ''
    },
    placeholder: {
      type: String,
      default: ''
    },
    /**
     * The display style, can be either dropdown, select, or button
     */
    appearance: {
      type: String,
      default: 'dropdown',
      validator: function (value) {
        return ['dropdown', 'select', 'button'].indexOf(value) !== -1
      }
    },
    /**
     * Override the text displayed on the button if `appearance` is `button` after an item
     * has been selected. By default display the selected item's label
     */
    buttonText: {
      type: String,
      default: ''
    },
    /**
     * Items are JSON objects with required 'label' and 'value'
     * {
     *  label: 'Item 1',
     *  value: 'item1'
     * }
     */
    items: {
      type: Array,
      required: false,
      default: () => [],
      // Items must have a label & value
      validator: (items) => !items.length || items.some(i => i.hasOwnProperty('label') && i.hasOwnProperty('value'))
    },
    /**
     * A flag to use fixed positioning of the popover to avoid content being clipped by parental boundaries.
     */
    positionFixed: {
      type: Boolean,
      default: false
    },
    /**
     * Override default filter functionality of case-insensitive search on label
     */
    filterFunc: {
      type: Function,
      default: ({ items, query }) => items.filter(item => item.label.toLowerCase().includes(query.toLowerCase()))
    },
    /**
     * Test mode - for testing only, strips out generated ids
     */
    testMode: {
      type: Boolean,
      default: false
    }
  },

  data: function () {
    return {
      filterStr: '',
      selectedItem: null,
      selectId: !this.testMode ? uuid.v1() : 'test-select-id-1234',
      selectInputId: !this.testMode ? uuid.v1() : 'test-select-input-id-1234',
      selectTextId: !this.testMode ? uuid.v1() : 'test-select-text-id-1234',
      selectItems: []
    }
  },

  computed: {
    boundKPopAttributes: function () {
      return {
        ...defaultKPopAttributes,
        ...this.kpopAttributes,
        popoverClasses: `${defaultKPopAttributes.popoverClasses} ${this.kpopAttributes.popoverClasses} k-select-pop-${this.appearance}`,
        width: this.width,
        disabled: this.$attrs.disabled
      }
    },
    listeners () {
      return this.$listeners
    },
    widthValue: function () {
      let w

      if (!this.width) {
        w = 200
        if (this.appearance === 'button') {
          w = 230
        }
      } else {
        w = this.width
      }

      return w === 'auto' ? w : w + 'px'
    },
    widthStyle: function () {
      return {
        width: this.widthValue
      }
    },
    filteredItems: function () {
      return this.filterFunc({ items: this.selectItems, query: this.filterStr })
    },
    placeholderText () {
      if (this.placeholder) {
        return this.placeholder
      } else if (this.$attrs.placeholder) {
        return this.$attrs.placeholder
      }

      if (this.appearance === 'button') {
        return 'Select an item'
      }

      return 'Filter...'
    },
    selectButtonText () {
      if (this.buttonText && this.selectedItem) {
        return this.buttonText
      } else if (this.selectedItem) {
        return this.selectedItem.label
      }

      return this.placeholderText
    }
  },
  beforeMount () {
    // items need keys
    this.selectItems = this.items

    for (let i = 0; i < this.selectItems.length; i++) {
      this.selectItems[i].key = `${this.selectItems[i].label.replace(' ', '-')}-${i}`
      if (this.selectItems[i].value === this.value || this.selectItems[i].selected) {
        this.selectItems[i].selected = true
        this.selectedItem = this.selectItems[i]
        this.selectItems[i].key += '-selected'

        if (this.appearance === 'select') {
          this.filterStr = this.selectedItem.label
        }
      }
    }
  },
  methods: {
    handleItemSelect (item) {
      this.selectItems.forEach(anItem => {
        if (anItem.key === item.key) {
          anItem.selected = true
          anItem.key += '-selected'
          this.selectedItem = anItem
        } else if (anItem.selected) {
          delete anItem.selected
          anItem.key = anItem.key.split('-selected')[0]
        }
      })
      this.filterStr = this.appearance === 'dropdown' ? '' : item.label
      this.$emit('selected', item)
      // this 'input' event must be emitted for v-model binding to work properly
      this.$emit('input', item.value)
      this.$emit('change', item)
    },
    clearSelection () {
      this.selectItems.forEach(anItem => {
        delete anItem.selected
        anItem.key = anItem.key.split('-selected')[0]
      })
      this.selectedItem = null
      // this 'input' event must be emitted for v-model binding to work properly
      this.$emit('input', null)
      this.$emit('change', null)
    },
    triggerFocus (isToggled) {
      const inputElem = document.getElementById(this.selectTextId)

      if (!isToggled && inputElem) { // simulate click to trigger dropdown open
        inputElem.click()
      }
    }
  }
}
</script>

<style lang="scss">
@import '~@kongponents/styles/variables';

.k-select {
  width: fit-content; // necessary for correct placement of popup

  .k-select-item-selection {
    background-color: var(--blue-100);
    color: var(--blue-500);
    font-weight: 400;
    display: flex;
    border-radius: 4px;
    margin-bottom: 6px;

    .selected-item-label {
      align-self: center;
      font-size: var(--type-xs);
      line-height: 16px;
    }

    .clear-selection-icon {
      margin-left: auto;
      margin-top: auto;
      margin-bottom: auto;
      padding: 0;
      height: 24px;

      .kong-icon {
        margin-left: auto;
      }
    }
  }

  .k-select-input {
    position: relative;
    display: inline-block;
    width: 100%;
    height: 32px;

    input.k-input {
      padding: var(--spacing-xs);
      height: 100%;
      border-radius: 4px 4px 0 0;
    }

    .kong-icon {
      position: absolute;
      top: 15px;
      right: 6px;
      z-index: 9;
    }

    .kong-icon-chevronDown {
      top: 10px;
    }
  }

  .k-select-button .has-caret .kong-icon {
    margin-left: auto;
  }

  .k-input {      // need this so input takes the
    width: 100%;  // k-input-wrapper's width which uses this.width prop
  }

  .k-select-trigger:after {
    display: inline-block;
    width: 0;
    height: 0;
    margin-left: var(--spacing-xs, spacing(xs));
    vertical-align: middle;
    content: "";
    border-top: 0.325em solid;
    border-right: 0.325em solid transparent;
    border-left: 0.325em solid transparent;
  }

  .k-select-popover {
    box-sizing: border-box;
    width: 100%;
    border-radius: 0 0 4px 4px;

    &.k-select-pop-button {
      --KPopPaddingY: var(--spacing-md);
      --KPopPaddingX: var(--spacing-xxs);
      border-radius: 4px;
      border: 1px solid var(--blue-200);
    }

    &.k-select-pop-dropdown {
      --KPopPaddingY: var(--spacing-md);
      --KPopPaddingX: var(--spacing-xxs);
      border: 1px solid var(--blue-200);
    }

    &.k-select-pop-select {
      --KPopPaddingY: var(--spacing-md);
      --KPopPaddingX: var(--spacing-xxs);
      border: 1px solid var(--black-10);
    }

    .k-select-empty-item button,
    .k-select-empty-item button:focus,
    .k-select-empty-item button:hover {
      font-style: italic;
      color: var(--grey-400);
    }

    ul {
      margin: 0;
      padding: 0;
    }

    a {
      flex: 1;
      color: var(--black-70);

      &:hover,
      &:active,
      &:focus {
        text-decoration: none;
      }
    }
  }
}
</style>
