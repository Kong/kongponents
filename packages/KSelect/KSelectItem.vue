<template>
  <div
    :key="item.key"
    :data-testid="`k-select-item-${item.value}`"
    class="k-select-item"
    @click="handleClick">
    <button
      :class="{ disabled, selected: item.selected }"
      :value="item.value">
      <slot>{{ item.label }}</slot>
      <KIcon
        v-if="item.selected"
        class="selected-item-icon"
        icon="check"
        color="var(--blue-200)" />
    </button>
  </div>
</template>

<script>
import KIcon from '@kongponents/kicon/KIcon.vue'

export default {
  name: 'SelectItem',
  components: { KIcon },
  props: {
    item: {
      type: Object,
      default: null,
      // Items must have a label and value
      validator: (item) => item.hasOwnProperty('label') && item.hasOwnProperty('value')
    },
    disabled: {
      type: Boolean,
      default: false
    }
  },
  methods: {
    handleClick () {
      this.$emit('selected', this.item)
    }
  }
}
</script>

<style lang="scss" scoped>
.k-select-item button {
  display: flex;
  align-items: center;
  font-size: var(--type-xs);
  line-height: 26px;
  color: var(--grey-500);
  padding-left: var(--spacing-xxs);
  width: 100%;
  border: none;
  font-family: var(--font-family-sans);
  background-color: var(--white);

  svg:not(.selected-item-icon) {
    margin-right: .75rem;
  }

  .selected-item-icon {
    margin-left: auto;
  }

  &:hover {
    background-color: var(--grey-100);
    color: var(--grey-600);
  }

  &.selected {
    background-color: var(--blue-100);
    color: var(--blue-500);
    font-weight: 400;
  }

  &.danger {
    color: var(--red-500);
  }
}
</style>
