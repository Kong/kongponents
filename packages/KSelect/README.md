# @kongponents/KSelect

[![](https://img.shields.io/npm/v/@kongponents/KSelect.svg?style=flat-square)](https://www.npmjs.com/package/@kongponents/KSelect)

```vue
<KSelect :description="'hello world'">
  Hello from a slot
</KSelect>
```
