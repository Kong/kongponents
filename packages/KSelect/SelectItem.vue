<template>
  <div>
    <li v-if="$scopedSlots.default && $listeners.click">
      <button
        :disabled="disabled"
        class="btn-link k-button non-visual-button"
        v-on="listeners" >
        <slot />
      </button>
    </li>
    <li
      v-else-if="$scopedSlots.default || (item && !item.to)"
      :class="{ disabled: disabled }">
      <slot>{{ item.label }}</slot>
    </li>
    <li v-else-if="item.to">
      <router-link
        :data-testid="item.label"
        :to="item.to">
        {{ item.label }}
      </router-link>
    </li>
  </div>
</template>

<script lang="ts">
export default {
  name: 'SelectItem',
  props: {
    item: {
      type: Object,
      required: false,
      default: null,
      // Items must have a label
      validator: (item) => item.hasOwnProperty('label')
    },
    disabled: {
      type: Boolean,
      default: false,
      required: false
    }
  }
}
</script>

<style lang="scss" scoped>
li {
  display: flex;
  align-items: center;
  font-size: var(--type-xs);
  line-height: 26px;
  color: var(--grey-600);

  svg {
    margin-right: .75rem;
  }

  &:hover {
    background-color: var(--grey-100);
  }

  button, a {
    text-align: left;
    padding: .625rem 1rem;
    text-decoration: none;
    width: 100%;
    color: var(--black-70);

    &:disabled {
      cursor: not-allowed;
    }
  }

  &.danger button:not(:disabled):hover {
    color: var(--black-70);
    transition: all 300ms;

    &:hover {
      color: var(--red-500);
    }
  }
}
</style>
