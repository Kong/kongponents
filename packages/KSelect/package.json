{
  "name": "@kongponents/kselect",
  "version": "6.1.1",
  "description": "Dropdown/Select component",
  "main": "dist/KSelect.umd.min.js",
  "componentName": "KSelect",
  "source": "KSelect.vue",
  "files": [
    "dist",
    "KSelect.vue"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@kongponents/kbutton": "^6.1.1",
    "@kongponents/kicon": "^6.1.1",
    "@kongponents/kinput": "^6.1.1",
    "@kongponents/klabel": "^6.1.1",
    "@kongponents/kpop": "^6.1.1",
    "@kongponents/ktoggle": "^6.1.1"
  }
}
