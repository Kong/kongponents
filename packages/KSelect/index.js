import KSelect from './KSelect'
import KSelectItem from './KSelectItem'

const Plugin = {
  install (Vue) {
    Vue.component('KSelect', KSelect)
    Vue.component('KSelectItem', KSelectItem)
  }
}

export default Plugin
export { KSelect, KSelectItem }

if (typeof window !== 'undefined' && window.Vue) {
  window.Vue.use(Plugin)
}
