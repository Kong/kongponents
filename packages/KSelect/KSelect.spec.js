import { mount } from '@vue/test-utils'
import KSelect from '@/KSelect/KSelect'

describe('KSelect', () => {
  it('matches snapshot', () => {
    const wrapper = mount(KSelect)

    expect(wrapper.html()).toMatchSnapshot()
  })
})
