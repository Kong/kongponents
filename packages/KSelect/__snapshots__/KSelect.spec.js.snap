// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KSelect matches snapshot 1`] = `
<div class="k-select" style="width: 200px;">
  <!---->
  <div id="test-select-id-1234" data-testid="k-select-selected-item">
    <!---->
    <div id="test-target-id-1234" aria-controls="test-popover-id-1234" role="button">
      <div id="test-select-input-id-1234" data-testid="k-select-input" role="listbox" class="" style="position: relative;">
        <!---->
        <div class="k-input-wrapper k-select-input"><input id="test-select-text-id-1234" placeholder="Filter..." class="form-control k-input k-input-medium">
          <!---->
          <!---->
        </div>
      </div>
      <div id="test-popover-id-1234" role="region" class="k-popover k-select-popover mt-0  k-select-pop-dropdown hide-caret" style="width: 0px; max-width: 0px; display: none;" name="fade">
        <!---->
        <div class="k-popover-content">
          <ul class="k-select-list ma-0 pa-0">
            <li data-testid="k-select-item-label1" class="k-select-item">
              <div role="option" class="d-block"><button value="label1" class=""><span class="k-select-item-label mr-2">Label 1</span> <span class="k-select-selected-icon-container"><!----></span></button></div>
            </li>
            <!---->
          </ul>
        </div>
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`KSelect renders props when passed 1`] = `
<div class="k-select" style="width: 200px;">
  <!---->
  <div id="test-select-id-1234" data-testid="k-select-selected-item">
    <!---->
    <div id="test-target-id-1234" aria-controls="test-popover-id-1234" role="button">
      <div id="test-select-input-id-1234" data-testid="k-select-input" role="listbox" class="" style="position: relative;">
        <!---->
        <div class="k-input-wrapper k-select-input"><input id="test-select-text-id-1234" placeholder="Filter..." class="form-control k-input k-input-medium">
          <!---->
          <!---->
        </div>
      </div>
      <div id="test-popover-id-1234" role="region" class="k-popover k-select-popover mt-0  k-select-pop-dropdown hide-caret" style="width: 0px; max-width: 0px; display: none;" name="fade">
        <!---->
        <div class="k-popover-content">
          <ul class="k-select-list ma-0 pa-0">
            <li data-testid="k-select-item-label1" class="k-select-item">
              <div role="option" class="d-block"><button value="label1" class=""><span class="k-select-item-label mr-2">Label 1</span> <span class="k-select-selected-icon-container"><!----></span></button></div>
            </li>
            <li data-testid="k-select-item-label2" class="k-select-item">
              <div role="option" class="d-block"><button value="label2" class=""><span class="k-select-item-label mr-2">Label 2</span> <span class="k-select-selected-icon-container"><!----></span></button></div>
            </li>
            <li data-testid="k-select-item-label3" class="k-select-item">
              <div role="option" class="d-block"><button value="label3" class=""><span class="k-select-item-label mr-2">Label 3</span> <span class="k-select-selected-icon-container"><!----></span></button></div>
            </li>
            <!---->
          </ul>
        </div>
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`KSelect renders with selected item 1`] = `
<div class="k-select" style="width: 200px;">
  <!---->
  <div id="test-select-id-1234" data-testid="k-select-selected-item">
    <div class="k-select-item-selection px-3 py-1">
      <div class="selected-item-label">Label 1</div> <button class="clear-selection-icon cursor-pointer non-visual-button"><span class="kong-icon kong-icon-clear"><svg viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
  <title>Clear</title>
  <circle cx="10" cy="10" r="6.25" fill="none" stroke="#A3BBCC" stroke-width="2" stroke-linecap="round"></circle>
  <path d="M12 12 8 8M12 8l-4 4" stroke="#A3BBCC" stroke-width="2" stroke-linecap="round"></path>
</svg>
</span></button>
    </div>
    <div id="test-target-id-1234" aria-controls="test-popover-id-1234" role="button">
      <div id="test-select-input-id-1234" data-testid="k-select-input" role="listbox" class="" style="position: relative;">
        <!---->
        <div class="k-input-wrapper k-select-input"><input id="test-select-text-id-1234" placeholder="Filter..." class="form-control k-input k-input-medium">
          <!---->
          <!---->
        </div>
      </div>
      <div id="test-popover-id-1234" role="region" class="k-popover k-select-popover mt-0  k-select-pop-dropdown hide-caret" style="width: 0px; max-width: 0px; display: none;" name="fade">
        <!---->
        <div class="k-popover-content">
          <ul class="k-select-list ma-0 pa-0">
            <li data-testid="k-select-item-label1" class="k-select-item">
              <div role="option" class="d-block"><button value="label1" class="selected"><span class="k-select-item-label mr-2">Label 1</span> <span class="k-select-selected-icon-container"><span class="kong-icon selected-item-icon kong-icon-check"><svg viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
  <path d="m14 7-6 6-3-3" fill="none" stroke="#A3BBCC" stroke-width="2.25" stroke-linecap="round" stroke-linejoin="round"></path>
</svg>
</span></span></button></div>
            </li>
            <!---->
          </ul>
        </div>
        <!---->
      </div>
    </div>
  </div>
</div>
`;
